consensus/core/src/tx.rs

```diff
@@ -267,21 +312,66 @@ impl<'a> ValidatedTransaction<'a> {
         assert!(tx.is_coinbase());
         Self { tx, entries: Vec::new(), calculated_fee: 0 }
     }
+}
 
-    pub fn populated_inputs(&self) -> impl ExactSizeIterator<Item = (&TransactionInput, &UtxoEntry)> {
-        self.tx.inputs.iter().zip(self.entries.iter())
+impl<'a> VerifiableTransaction for ValidatedTransaction<'a> {
+    fn tx(&self) -> &Transaction {
+        self.tx
     }
 
-    pub fn outputs(&self) -> &[TransactionOutput] {
-        &self.tx.outputs
+    fn populated_input(&self, index: usize) -> (&TransactionInput, &UtxoEntry) {
+        (&self.tx.inputs[index], &self.entries[index])
     }
+}
 
-    pub fn is_coinbase(&self) -> bool {
-        self.tx.is_coinbase()
+/// Represents a mutable owned transaction along with partially filled UTXO entry data and optional fee and mass
+pub struct MutableTransaction {
+    /// The inner transaction
+    pub tx: Transaction,
+    /// Partially filled UTXO entry data
+    pub entries: Vec<Option<UtxoEntry>>,
+    /// Populated fee
+    pub calculated_fee: Option<u64>,
+    /// Populated mass
+    pub calculated_mass: Option<u64>,
```

Comment:
Maybe it should be private? Calculated based on entries, and recalculate each time you add/remove an entry?

Reply:
AFAIU, calculated mass is based on the inner `Transaction` outputs. It does not depend on the entries.

Reply:
Fee calc only makes sense once all utxo entries are obtained, and never changes after. Mass requires the mass calculator bcs it's based on consensus params